src/Hedgehog/GenConfig.fs

```diff
@@ -0,0 +1,16 @@
+namespace Hedgehog
+
+type GenConfig<'a> = private {
+    Formatter : 'a -> string
+}
+
+module GenConfig =
+
+    let defaultConfig () =
+        { Formatter = sprintf "%A" }
```

Comment:
For me, when I see something with the signature `unit -> 'T`, I think it side effects somewhere. Since this doesn't, would it be reasonable to drop `()`?

Reply:
This is running into F# compiler's inability to infer generic types for values.  The easiest workaround, as was done there, is to change it from a value to a function (by accepting the value `()`).  Another workaround is to explicitly specify the type parameters (so that the F# compiler doesn't have to do any inference).  I will give one possibility in a different comment thread (that includes line 9).

Reply:
> I will give one possibility in a different comment thread (that includes line 9).

Actually, that is not possible.  Either a bug or missing feature in GitHub is that a new multiline comment thread cannot be made if the last line already has a comment thread.

Here is one way to explicitly specify the type parameters.  Maybe there is a way to do this inline, but I don't know of such code at the moment.

```fsharp
    let defaultConfig<'a> =
        let f : 'a -> string = sprintf "%A"
        { Formatter = f }
```

Reply:
This works:

```f#
let f : 'a -> string = sprintf "%A" in { Formatter = f }
```

I'm not necessarily advocating for it.

Reply:
> This is running into F# compiler's inability to infer generic types for values.

I know what you mean, but this is _slightly_ wrong. The F# compiler can't emit generic values because support for this doesn't exist in MSIL. The equivalent in C# would be:

```csharp
public GenConfig<T> DefaultConfig<T> = new GenConfig<T> {
    Formatter = // ...
}
```

Reply:
What did I say that is wrong?

Reply:
> This is running into F# compiler's inability to infer generic types for values.
> What did I say that is wrong?

It can infer the types just fine, the resulting construct can't be represented in MSIL.

Reply:
> I'm not necessarily advocating for it.

To be clear - I still am advocating for the removal of `unit` as a param. This will also have better symmetry with `PropertyConfig.defaultConfig`.

Reply:
Do you think it is possible for the F# compiler to be changed so that
```fsharp
let defaultConfig = { Formatter = sprintf "%A" }
```

and
```fsharp
let defaultConfig = { Formatter = let f : 'a -> string = sprintf "%A" in { Formatter = f } }
```

emit the same MSIL?

Reply:
I tried your examples (all permutations I could think of) and still got errors..

```fsharp
type Blah<'a> = { Formatter : 'a -> string }

let defaultConfig =
    { Formatter = let f : 'a -> string = sprintf "%A" in { Formatter = f } }
// This expression was expected to have type
//     ''a -> string'
// but here has type
//     'Blah<'b>'

let defaultConfig =
    { Formatter = let f : 'a -> string = sprintf "%A" in f }
// Value restriction. The value 'defaultConfig' has been inferred to have generic type
//     val internal defaultConfig : Blah<'_a>    
// Either define 'defaultConfig' as a simple data term, make it a function with explicit arguments or, if you do not intend for it to be generic, add a type annotation.

let defaultConfig =
    let f : 'a -> string = sprintf "%A" in { Formatter = f }
// Value restriction. The value 'defaultConfig' has been inferred to have generic type
//     val internal defaultConfig : Blah<'_a>    
// Either define 'defaultConfig' as a simple data term, make it a function with explicit arguments or, if you do not intend for it to be generic, add a type annotation.
```

> emit the same MSIL?

I only now realized I've been misspeaking. MSIL is the bytecode that is JIT'd at runtime to a host architecture. That's not what I meant, I meant the binary format of .NET assemblies. Fields/Properties can't be declared as generic. In theory properties _could_ support this since they themselves are just methods. However, the .NET teams have regularly issued resistance to any binary format changes, understandably so.

Reply:
> I tried your examples (all permutations I could think of) and still got errors..

You forgot to add an explicit type parameter on the value.

Some of our code blocks above also forgot to add the explicit type parameter is some cases.  In comment https://github.com/hedgehogqa/fsharp-hedgehog/pull/353#discussion_r711604381, which was my first comment with a code block in this thread, I did not forget the explicit type parameter.

Reply:
> Fields/Properties can't be declared as generic.

I would typically interpret this to mean "the type" of a field/property cannot be generic.  Of course that statement is false.  I think what you mean is that in both C# and F#, neither a field nor a property can "pick up" or "introduce" a new type parameter when referenced.  I think that statement is true.

> Do you think it is possible for the F# compiler to be changed so that
> 
> ```fsharp
> let defaultConfig = { Formatter = sprintf "%A" }
> ```
> 
> and
> 
> ```fsharp
> let defaultConfig<'a> = let f : 'a -> string = sprintf "%A" in { Formatter = f }
> ```
> 
> emit the same MSIL?

(I fixed two typos in the second code block compared to what I am quoting from https://github.com/hedgehogqa/fsharp-hedgehog/pull/353#discussion_r711622977)

I am not suggesting that a new type parameter be introduced when `defaultConfig` is accessed. I am suggesting that it is possible that F# to be changed/improved so it correctly infers the use of an already existing type parameter (instead of resolving that type parameter as `obj` like it does in its current behavior).

I think it should be possible, but I don't know for sure.  Maybe I am wrong about something here.

Reply:
It can already infer the types, H&M type inference is pretty solid. F# just doesn't allow this because the CLR doesn't. The best you can do today as an F# programmer is make default config an inline function to remove the overhead of the call (but you then introduce/keep instantiation overhead).

In a perfect world F# could "memoize" copies of this property for each type argument its been given, and pretend the field is generic. Such that `defaultConfig<int32>` is only created once, as is `defaultConfig<string list>`, etc. While this is theoretically possible, I don't think it would ever happen without CLR support first.

Does that clear things up at all?

Reply:
> Does that clear things up at all?

No, you are wrong.

> The best you can do today as an F# programmer is make default config an inline function to remove the overhead of the call [...]

This best one can do is what I gave in https://github.com/hedgehogqa/fsharp-hedgehog/pull/353#discussion_r711604381.

> F# just doesn't allow this because the CLR doesn't.

The code in the linked comment compiles and does what we want.  I don't care what CLI it generates.  F# can do things that the CLI doesn't support, such as statically resolved type parameters and units of measure.  Maybe this can't be accessed from C#.  That is a different issue.  I think we should first give F# the best possible API and then consider what can be done for C#.